modules/core/src/main/java/org/apache/ignite/internal/util/nio/GridNioServer.java

```diff
@@ -2577,21 +2568,33 @@ private void processSelectedKeys(Set<SelectionKey> keys) throws ClosedByInterrup
                     throw e;
                 }
                 catch (Exception | Error e) { // TODO IGNITE-2659.
-                    try {
-                        U.sleep(1000);
-                    }
-                    catch (IgniteInterruptedCheckedException ignore) {
-                        // No-op.
-                    }
-
-                    GridSelectorNioSessionImpl ses = attach.session();
+                    procesKeysSelectionError(e, attach);
+                }
+            }
+        }
 
-                    if (!closed)
-                        U.error(log, "Failed to process selector key [ses=" + ses + ']', e);
-                    else if (log.isDebugEnabled())
-                        log.debug("Failed to process selector key [ses=" + ses + ", err=" + e + ']');
+        /**
+         * Processes errors occured while processing selector keys.
+         *
+         * @param e Exception or error which occured.
+         * @param attach GridNioKeyAttachment.
+         */
+        private void procesKeysSelectionError(Throwable e, GridNioKeyAttachment attach) {
+            if (!X.hasCause(e, Exception.class)) {
```

Comment:
If we think here only about Assertion or other Errors, change the condition to _X.hasCause(e, Error.class)_.
But it only my view.